File: python/air temp/5.py
# 2024 vs 2021 air temp 4 figures    
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from matplotlib.dates import DateFormatter

file_path = "данныепроекта.xlsx"

df = pd.read_excel(file_path, header=0)

df = df.rename(columns={
    'Регион': 'region',
    'Станция_айди': 'station_id',
    'Станция': 'station_name',
    'Дата': 'date',
    'Сред': 't_mean',
    'Макс': 't_max',
    'Мин': 't_min'
})

df = df.dropna(how='all')

df['date'] = pd.to_datetime(df['date'], format='%d.%m.%Y', errors='coerce')
df = df.dropna(subset=['date'])

df['t_mean'] = pd.to_numeric(df['t_mean'], errors='coerce')
df['t_max'] = pd.to_numeric(df['t_max'], errors='coerce')
df['t_min'] = pd.to_numeric(df['t_min'], errors='coerce')

df['year'] = df['date'].dt.year
df['month'] = df['date'].dt.month

df = df[df['month'].isin([2, 3, 4])]

# 1. ГРАФИК РАЗНИЦЫ (2024 - 2021)
df24 = df[df['year'] == 2024]
df21 = df[df['year'] == 2021]

reg24 = df24.groupby(['region', 'date'])['t_mean'].mean().reset_index()
reg21 = df21.groupby(['region', 'date'])['t_mean'].mean().reset_index()

min_len = min(len(reg24), len(reg21))
reg24 = reg24.iloc[:min_len]
reg21 = reg21.iloc[:min_len]

reg_diff = reg24.copy()
reg_diff['t_diff'] = reg24['t_mean'].values - reg21['t_mean'].values

plt.figure(figsize=(14, 10))
for region in reg_diff['region'].unique():
    sub = reg_diff[reg_diff['region'] == region]
    plt.plot(sub['date'], sub['t_diff'], label=region)

plt.axhline(0, color='black', linestyle='--')
plt.title("Разница средней температуры (2024 – 2021)", fontsize=16, fontweight='bold')
plt.xlabel("Дата")
plt.ylabel("ΔT, °C")
plt.legend()
plt.grid(True)
plt.xticks(rotation=45)
plt.savefig('raznica2024vs2021')
plt.show()

# 2. СКОРОСТЬ ПРОГРЕВА (dT/dt)
df24s = df24.sort_values(['station_id', 'date'])
df21s = df21.sort_values(['station_id', 'date'])

df24s['dT_dt'] = df24s.groupby('station_id')['t_mean'].diff()
df21s['dT_dt'] = df21s.groupby('station_id')['t_mean'].diff()

dT24 = df24s.groupby('date')['dT_dt'].mean().reset_index()
dT21 = df21s.groupby('date')['dT_dt'].mean().reset_index()

plt.figure(figsize=(14, 10))
plt.plot(dT24['date'], dT24['dT_dt'], label="2024", color="red")
plt.plot(dT21['date'], dT21['dT_dt'], label="2021", color="blue")
plt.axhline(0, color='black', linestyle='--')
plt.title("Скорость прогрева воздуха (dT/dt): 2024 vs 2021", fontsize=16, fontweight='bold')
plt.xlabel("Дата")
plt.ylabel("dT/dt (°C/день)")
plt.legend()
plt.grid(True)
plt.xticks(rotation=45)
plt.savefig('scorostprogreva2024vs2021')
plt.show()

# 3. График начала таяния (точки)
def thaw_day(st_df):
    pos = st_df[st_df['t_mean'] > 0]
    if len(pos) == 0:
        return None
    return pos['date'].iloc[0]

thaw = df.groupby(['year', 'region', 'station_id']).apply(thaw_day).reset_index()
thaw.columns = ['year', 'region', 'station_id', 'first_thaw']

plt.figure(figsize=(14, 10))
for yr in thaw['year'].unique():
    sub = thaw[thaw['year'] == yr]
    plt.scatter(sub['region'], sub['first_thaw'], label=str(yr))

plt.title("Первая дата перехода температуры > 0°C по станциям", fontsize=16, fontweight='bold')
plt.xlabel("Регион")
plt.ylabel("Дата")
plt.legend()
plt.grid(True)
plt.xticks(rotation=45)
plt.savefig('nachalotayania2024vs2021')
plt.show()

# 4. HEATMAP сравнения месяцев
monthly = df.groupby(['year', 'region', 'month'])['t_mean'].mean().reset_index()
heat = monthly.pivot_table(values='t_mean', index='region', columns=['year', 'month'])

plt.figure(figsize=(12, 10))
plt.imshow(heat, cmap='coolwarm', aspect='auto')
plt.title("Heatmap: средняя температура по месяцам, регионам и годам", fontsize=16, fontweight='bold')
plt.colorbar(label='Температура, °C')
plt.xticks(range(len(heat.columns)), heat.columns, rotation=90)
plt.yticks(range(len(heat.index)), heat.index)
plt.savefig('heatmap2024vs2021')
plt.show()

File: matlab/air temp/5.m
%% 2024 vs 2021 AIR TEMPERATURE ANALYSIS — MATLAB VERSION 
clc; clear; close all;

file_path = "данныепроекта.xlsx";

opts = detectImportOptions(file_path, 'VariableNamingRule','preserve');
T = readtable(file_path, opts);

T.Properties.VariableNames = { ...
    'region','station_id','station_name','date','t_mean','t_max','t_min'};

allEmpty = all(ismissing(T), 2);
T(allEmpty,:) = [];

T.date = datetime(T.date,'InputFormat','dd.MM.yyyy','Format','dd.MM.yyyy');

numVars = ["t_mean","t_max","t_min"];
for v = numVars
    T.(v) = str2double(string(T.(v)));
end

T.year  = year(T.date);
T.month = month(T.date);

T = T(ismember(T.month,[2 3 4]), :);

T.region = string(T.region);


%% TEMPERATURE DIFFERENCE: 2024 – 2021

T24 = T(T.year==2024,:);
T21 = T(T.year==2021,:);

reg24 = groupsummary(T24, ["region","date"], "mean", "t_mean");
reg21 = groupsummary(T21, ["region","date"], "mean", "t_mean");

min_len = min(height(reg24), height(reg21));
reg24 = reg24(1:min_len,:);
reg21 = reg21(1:min_len,:);

reg_diff = table();
reg_diff.region = reg24.region;
reg_diff.date   = reg24.date;
reg_diff.t_diff = reg24.mean_t_mean - reg21.mean_t_mean;

% Plot ΔT
figure('Position',[100 100 1200 700]); hold on;
regions = unique(reg_diff.region);

for r = regions'
    mask = reg_diff.region == r;
    plot(reg_diff.date(mask), reg_diff.t_diff(mask), 'LineWidth',1.6, ...
        'DisplayName', r);
end

yline(0,'--k','HandleVisibility','off');
xlabel("Date");
ylabel("\Delta Temperature (°C)");
% title("Difference in Mean Temperature (2024 - 2021)");
legend show; grid on; xtickangle(45);
print('raznica2024vs2021','-dpng','-r300');


%% dT/dt 
% --- 2024 ---
T24s = sortrows(T24, ["station_id","date"]);
G24 = findgroups(T24s.station_id);
d24_cell = splitapply(@(x){[NaN; diff(x)]}, T24s.t_mean, G24);
T24s.dT = vertcat(d24_cell{:});

% --- 2021 ---
T21s = sortrows(T21, ["station_id","date"]);
G21 = findgroups(T21s.station_id);
d21_cell = splitapply(@(x){[NaN; diff(x)]}, T21s.t_mean, G21);
T21s.dT = vertcat(d21_cell{:});

% Average dT/dt by date
dT24 = groupsummary(T24s, "date", "mean", "dT");
dT21 = groupsummary(T21s, "date", "mean", "dT");

figure('Position',[100 100 1200 700]); hold on;
plot(dT24.date, dT24.mean_dT, 'r','LineWidth',1.6,'DisplayName',"2024");
plot(dT21.date, dT21.mean_dT, 'b','LineWidth',1.6,'DisplayName',"2021");
yline(0,'--k','HandleVisibility','off');
xlabel("Date");
ylabel("dT/dt (°C/day)");
% title("Heating Rate (dT/dt): 2024 vs 2021");
legend show; grid on; xtickangle(45);
print('scorostprogreva2024vs2021','-dpng','-r300');


%% FIRST THAW DATE 

% Thaw function (first t_mean > 0)
find_thaw = @(dates, temps) firstThawFunc(dates, temps);

function out = firstThawFunc(dates, temps)
    mask = temps > 0;
    if any(mask)
        out = dates(find(mask,1,'first'));
    else
        out = NaT;
    end
end

[G, year_g, region_g, station_g] = findgroups(T.year, T.region, T.station_id);

first_thaw_vec = splitapply(@(d,t) find_thaw(d,t), ...
    T.date, T.t_mean, G);

thaw = table(year_g, region_g, station_g, first_thaw_vec, ...
    'VariableNames',{'year','region','station_id','first_thaw'});

figure('Position',[100 100 1200 700]); hold on;

for yr = unique(thaw.year)'
    mask = thaw.year == yr;

    reg_cat = categorical(thaw.region(mask));   % X-axis fix

    scatter(reg_cat, thaw.first_thaw(mask), ...
        60, 'filled', 'DisplayName', string(yr));
end

xlabel("Region");
ylabel("First Day Temperature > 0°C");
% title("First Thaw Date by Station");
legend show; grid on; xtickangle(45);
print('nachalotayania2024vs2021','-dpng','-r300');


%% HEATMAP

monthly = groupsummary(T, ["year","region","month"], "mean", "t_mean");

regions = unique(monthly.region);
years = unique(monthly.year);
months = unique(monthly.month);

colNames = strings(1, numel(years)*numel(months));
heat = NaN(numel(regions), numel(colNames));

idx = 1;
for y = years'
    for m = months'
        colNames(idx) = sprintf("%d_%02d", y, m);
        for r = 1:numel(regions)
            mask = monthly.year==y & monthly.month==m & monthly.region==regions(r);
            row = monthly(mask,:);
            if ~isempty(row)
                heat(r,idx) = row.mean_t_mean;
            end
        end
        idx = idx + 1;
    end
end

figure('Position',[100 100 1200 800]);
imagesc(heat);
colormap('turbo');

cb = colorbar;
cb.Label.String = 'Mean Temperature (°C)';
cb.Label.FontSize = 12;
cb.Label.FontWeight = 'bold';

minT = min(heat(:), [], 'omitnan');
maxT = max(heat(:), [], 'omitnan');
midT = (minT + maxT) / 2;

cb.Ticks = [minT midT maxT];
cb.TickLabels = {
    sprintf('%.1f°C (min)', minT)
    sprintf('%.1f°C (mid)', midT)
    sprintf('%.1f°C (max)', maxT)
};

caxis([minT maxT]);

ylabel("Region");
xlabel("Year–Month");

set(gca, 'YTick', 1:numel(regions), 'YTickLabel', regions);
set(gca, 'XTick', 1:numel(colNames), 'XTickLabel', colNames);
xtickangle(90);

% title("Heatmap: Mean Temperature by Region and Month");
print('heatmap2024vs2021','-dpng','-r300');
